File: python-sdk/amadeus_sdk/rag.py
"""RAG management — ingest, query, list, and delete documents via the API."""

from __future__ import annotations

from typing import Optional

from .client import AmadeusClient
from .types import RagDocumentInfo, RagIngestResponse, RagQueryResponse, RagSearchResult


class RAGManager:
    """Manage RAG document ingestion and semantic search via the API.

    Usage::

        rag = RAGManager(client)
        await rag.ingest_text("Amadeus is a Rust AI agent SDK.", doc_id="about")
        results = await rag.query("What language is Amadeus written in?")
        docs = await rag.list_documents()
        await rag.delete_document("about")
    """

    def __init__(self, client: AmadeusClient) -> None:
        self._client = client

    async def ingest_text(
        self,
        text: str,
        document_id: Optional[str] = None,
        chunk_size: Optional[int] = None,
        chunk_overlap: Optional[int] = None,
    ) -> RagIngestResponse:
        """Ingest raw text into the vector store.

        Args:
            text: The text content to chunk and embed.
            document_id: Optional human-readable ID. Auto-generated if omitted.
            chunk_size: Target characters per chunk.
            chunk_overlap: Overlap characters between adjacent chunks.

        Returns:
            RagIngestResponse with document_id and chunk_count.
        """
        return await self._client.rag_ingest(
            text=text,
            document_id=document_id,
            chunk_size=chunk_size,
            chunk_overlap=chunk_overlap,
        )

    async def ingest_file(
        self,
        path: str,
        document_id: Optional[str] = None,
        chunk_size: Optional[int] = None,
        chunk_overlap: Optional[int] = None,
    ) -> RagIngestResponse:
        """Ingest a local file into the vector store.

        Args:
            path: Path to the file to ingest.
            document_id: Optional human-readable ID. Auto-generated if omitted.
            chunk_size: Target characters per chunk.
            chunk_overlap: Overlap characters between adjacent chunks.

        Returns:
            RagIngestResponse with document_id and chunk_count.
        """
        return await self._client.rag_ingest(
            path=path,
            document_id=document_id,
            chunk_size=chunk_size,
            chunk_overlap=chunk_overlap,
        )

    async def query(self, query: str, top_k: int = 5) -> list[RagSearchResult]:
        """Semantic search over ingested documents.

        Args:
            query: Natural language search query.
            top_k: Number of top results to return.

        Returns:
            List of RagSearchResult with content and relevance scores.
        """
        resp = await self._client.rag_query(query, top_k=top_k)
        return resp.results

    async def list_documents(self) -> list[RagDocumentInfo]:
        """List all ingested documents with chunk counts.

        Returns:
            List of RagDocumentInfo.
        """
        resp = await self._client.rag_list_documents()
        return resp.documents

    async def delete_document(self, document_id: str) -> dict:
        """Delete a document and all its chunks.

        Args:
            document_id: The document ID to delete.

        Returns:
            API response dict with deleted count and document_id.
        """
        return await self._client.rag_delete_document(document_id)

File: crates/api/src/api/types/rag.rs
// @amadeus-header
// summary: Retrieval ingestion, query, result, and document API data types.
// layer: api
// status: active
// feature_flags:
// - api
// provides:
// - type: crate::api::types::RagDocumentsResponse
// - type: crate::api::types::RagIngestRequest
// - type: crate::api::types::RagQueryRequest
// - type: crate::api::types::RagQueryResponse
// uses:
// - protocol: serde serialization
// invariants:
// - Serialized fields retain the retrieval HTTP contract.
// side_effects: none
// tests:
// - cmd: cargo test -p api --all-features
// @end-amadeus-header

use serde::{Deserialize, Serialize};

// ---------------------------------------------------------------------------
// RAG API types
// ---------------------------------------------------------------------------

/// Request for `POST /rag/ingest`.
#[derive(Debug, Deserialize)]
pub struct RagIngestRequest {
    /// Raw text to ingest directly.
    #[serde(default)]
    pub text: Option<String>,
    /// Path to a local file to ingest.
    #[serde(default)]
    pub path: Option<String>,
    /// Optional human-readable document identifier. Auto-generated if not provided.
    #[serde(default)]
    pub document_id: Option<String>,
    /// Target characters per chunk.
    #[serde(default)]
    pub chunk_size: Option<usize>,
    /// Overlap characters between adjacent chunks.
    #[serde(default)]
    pub chunk_overlap: Option<usize>,
}

/// Response for `POST /rag/ingest`.
#[derive(Debug, Serialize)]
pub struct RagIngestResponse {
    pub document_id: String,
    pub chunk_count: usize,
}

/// Request for `POST /rag/query`.
#[derive(Debug, Deserialize)]
pub struct RagQueryRequest {
    /// Natural language query for semantic search.
    pub query: String,
    /// Number of top results to return (default from config).
    #[serde(default)]
    pub top_k: Option<usize>,
}

/// Response for `POST /rag/query`.
#[derive(Debug, Serialize)]
pub struct RagQueryResponse {
    pub query: String,
    pub results: Vec<RagSearchResult>,
}

/// A single semantic search result with relevance score.
#[derive(Debug, Serialize)]
pub struct RagSearchResult {
    pub rank: usize,
    pub key: String,
    pub content: String,
    pub source: String,
    pub score: f32,
}

/// Response for `GET /rag/documents`.
#[derive(Debug, Serialize)]
pub struct RagDocumentsResponse {
    pub documents: Vec<RagDocumentInfo>,
}

/// Summary info for an ingested document.
#[derive(Debug, Serialize)]
pub struct RagDocumentInfo {
    pub id: String,
    pub chunk_count: usize,
    pub ingested_at: String,
}
